<!-- Launch file for the rrt-detector and the assigner -->
<launch>
  <!-- argument need to change for deployment-->
  <arg name="global_map" value ="/merged/map_processed"/>
  <arg name="global_frame" value ="world"/>
  <arg name="global_costmap_topic" value="/move_base/global_costmap/costmap"/>
  <arg name="plan_service" value="/move_base/NavfnROS/make_plan"/>
  <arg name="robot_base_link" value="base_link"/>
  <arg name="move_base_service" value="/move_base"/>
  <arg name="time_per_meter" value="6"/>

  <!-- robot name space definition -->
  <arg name="first_robot"  default="multijackal_01"/>
  <arg name="second_robot" default="multijackal_03"/>
  <!-- <arg name="first_robot"  default="multijackal_02"/>
  <arg name="second_robot" default="multijackal_02"/> -->

  <!-- RRT common parameters -->
  <arg name="info_radius" value="1.2"/>
  <arg name="info_multiplier" value="3.0"/>
  <arg name="hysteresis_radius" value="5.0"/>
  <arg name="hysteresis_gain" value="5.0"/>
  <arg name="costmap_clearing_threshold" value="50"/>
  <!-- <arg name="delay_after_assignement" value="2.4"/> -->
  <arg name="delay_after_assignement" value="2.4"/>
  <arg name="goal_time_thres" value="6.0"/>
  <arg name="frequency1" value="5"/>
  <arg name="frequency2" value="10"/>
  <arg name="bandwith_cluster" value="2.5"/>
  <arg name="detected_frontier" value="/detected_points"/>
  <arg name="filtered_frontier" value="/filtered_points"/>
  <arg name="rd_metric" value="25.0"/>


  <!-- manual assignment of the exploration starting position -->
  <arg name="initialStartPts" value ="2.32,2.78"/>
  <arg name="odom_topic_startpts" value="/tb3_1/odom"/>
  <arg name="robot_frame" value="/tb3_1/base_link"/>

  <!-- for controlling the RRT growth distance -->
  <arg name="Geta" value="2.0"/>
  
  <!-- boundary node for checking clicked point and time -->
  <node pkg="rrt_exploration" type="boundary.py" name="exploration_boundary" output="screen">
    <param name="mapFrame" value="$(arg global_frame)"/>
    <param name="n_point" value="4"/>
    <param name="topicInput" value="/clicked_point"/>
    <param name="topicOutput" value="/exploration_boundary"/>
    <param name="frequency" value="0.5"/>
    <param name="mapTopic" value="/map"/>
    <!-- <param name="initialStartPts" value="$(arg initialStartPts)"/> -->
    <param name="odom_topic_startpts" value="$(arg odom_topic_startpts)"/>
    <param name="robot_frame" value="$(arg robot_frame)"/>

    <remap from="/tf" to="/merged_desktop/tf" />
    <remap from="/tf_static" to="/merged_desktop/tf_static" />
  </node>

  <!-- global RRT detector -->
  <node pkg="rrt_exploration" type="global_rrt_detector" name="global_rrt_detector" output="screen">
    <param name="eta" value="$(arg Geta)"/>
    <param name="map_topic" value="$(arg global_map)"/>

    <remap from="/tf" to="/merged_desktop/tf" />
    <remap from="/tf_static" to="/merged_desktop/tf_static" />
  </node>

  <!-- filtering node -->
    <node pkg="rrt_exploration" type="filter.py" name="filter" output="screen">
      <param name="map_topic" value="$(arg global_map)"/>
      <param name="info_radius" value="$(arg info_radius)"/>
      <param name="costmap_clearing_threshold" value="$(arg costmap_clearing_threshold)"/>
      <param name="global_costmap_topic" value="$(arg global_costmap_topic)"/>
      <param name="goals_topic" value="$(arg detected_frontier)"/>
      <param name="robot_namelist" value="$(arg first_robot),$(arg second_robot)"/>
      <param name="rate" value="$(arg frequency1)"/>
      <param name="local_map" value="/map"/>
      <param name="bandwith_cluster" value="$(arg bandwith_cluster)"/>

      <remap from="/tf" to="/merged_desktop/tf" />
      <remap from="/tf_static" to="/merged_desktop/tf_static" />
    </node>


  <!-- assigner node -->
  <node pkg="rrt_exploration" type="assigner.py" name="assigner" output="screen">
    <param name="map_topic" value="$(arg global_map)"/>
    <param name="global_frame" value="$(arg global_frame)"/>
    <param name="info_radius" value="$(arg info_radius)"/>
    <param name="info_multiplier" value="$(arg info_multiplier)"/>
    <param name="hysteresis_radius" value="$(arg hysteresis_radius)"/>
    <param name="hysteresis_gain" value="$(arg hysteresis_gain)"/>
    <param name="frontiers_topic" value="$(arg filtered_frontier)"/>
    <param name="robot_namelist" value="$(arg first_robot),$(arg second_robot)"/>
    <param name="time_per_meter" value="$(arg time_per_meter)"/>
    <param name="delay_after_assignement" value="$(arg delay_after_assignement)"/>
    <param name="rate" value="$(arg frequency2)"/>
    <param name="plan_service" value="$(arg plan_service)"/>
    <param name="base_link" value="$(arg robot_base_link)"/>
    <param name="move_base_service" value="$(arg move_base_service)"/>
    <param name="rd_metric" value="$(arg rd_metric)"/>
    <param name="local_map" value="map"/>

    <remap from="/tf" to="/merged_desktop/tf" />
    <remap from="/tf_static" to="/merged_desktop/tf_static" />
  </node>
</launch>
